text/0037-node-upgrades.md

```diff
@@ -0,0 +1,98 @@
+Information cached includes which mode of operation the node was executed in, the protocol version, the path to the node
+binary and the path to the config folder.
+
+On a fresh start, the launcher searches for the lowest installed version of `casper-node` and starts running it in
+validator mode.
+
+If the launcher cannot find an appropriate version at any stage of upgrading or downgrading, it exits with an error.
+
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+### casper-node-launcher
+
+The following describes the control flow in the launcher:
+
+![control flow in the launcher](images/0037/casper-launcher-flow.png "control flow in the launcher")
+
+### casper-node
+
+On startup, the node needs to decide what action to take depending upon whether it is an initial run (without fast sync
+this will always be treated as running from genesis), a restart after an upgrade, or a restart after an unexpected
+shutdown.  The previous run (if any) is deduced by examining the stored linear chain, specifically by querying the
+highest block, and comparing that to the current and next (if available) activation points.
+
+Note that if a node starts with a chainspec which specifies `protocol.hard_reset = true`, then all blocks created in or
+after the chainspec's activation point are effectively removed from storage, creating a situation as if the node had
+ceased operating at the end of the appropriate era.  This option is seen as a failsafe, to cater for the scenario where
+the network has ceased operating properly mid-era.
+
+The actions a node will take on startup, in the order in which they are tested, are summarized below:
+
+* HB = highest block
+* PV = protocol version
+* AP = activation point
+* current X = value taken from current chainspec
+* next X = value taken from next upgraded chainspec
+
+| Condition                                           | Scenario                                     | Action                                   |
+| :-------------------------------------------------- | :------------------------------------------- | :--------------------------------------- |
+| no blocks in storage, and PV 1.0.0                  | initial run                                  | commit genesis and continue running      |
+| no blocks in storage, and PV not 1.0.0              | invalid initial run                          | exit with code 102 to downgrade          |
+| HB.era + 1 == current AP and HP is switch block     | restart immediately after upgrade            | commit upgrade and continue running      |
+| HB.era + 1 == current AP and HP is not switch block | invalid run - blocks missing from storage    | exit with code 102 to downgrade          |
+| HB.era < current AP                                 | invalid run - blocks missing from storage    | exit with code 102 to downgrade          |
```

Comment:
```suggestion
| HB.era + 1 < current AP                             | invalid run - blocks missing from storage    | exit with code 102 to downgrade          |
```
(I guess that's equivalent because of the order, but it might still be clearer.)

Reply:
Fixed in [d62ffc1](https://github.com/CasperLabs/ceps/pull/37/commits/d62ffc1d930e7d2dc6d5d14659e13f144b5b4308).